### PR TITLE
Save Jitsi allowed domain in account data

### PR DESCRIPTION
**Pull request by BillCarsonFr**
### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request updates [CHANGES.rst](https://github.com/matrix-org/matrix-android-sdk/blob/develop/CHANGES.rst)

### CODE REVIEW

File: matrix-sdk/src/main/java/org/matrix/androidsdk/features/integrationmanager/IntegrationManager.kt [hunk 1]
```diff
@@ -106,17 +113,66 @@ class IntegrationManager(val mxSession: MXSession, val context: Context) {
...
+        widgetPermissions =
+                // Remove existing perm for current widget if any, then add updated state
+                widgetPermissions.filterNot { it.stateEventId == stateEventId } + listOf(WidgetPermission(stateEventId, allowed))
+        notifyListeners()
+
+        mxSession.accountDataRestClient.setAccountData(
+                mxSession.myUserId,
+                AccountDataElement.ACCOUNT_DATA_TYPE_ALLOWED_WIDGETS,
+                updatedMap,
+                object : SimpleApiCallback<Void>(callback) {
```
Comment (bmarty) [inline matrix-sdk/src/main/java/org/matrix/androidsdk/features/integrationmanager/IntegrationManager.kt]: Just pass the callback then, no need to wrap :)

File: matrix-sdk/src/main/java/org/matrix/androidsdk/features/integrationmanager/IntegrationManager.kt [hunk 2]
```diff
@@ -80,13 +86,14 @@ class IntegrationManager(val mxSession: MXSession, val context: Context) {
...
 
     fun enableIntegrationManagerUsage(allowed: Boolean, callback: ApiCallback<Void>) {
+        // Optimistic update before account data sync
+        if (integrationAllowed != allowed) {
+            integrationAllowed = allowed
+            notifyListeners()
+        }
+        
         mxSession.enableIntegrationManagerUsage(allowed, object : SimpleApiCallback<Void>(callback) {
             override fun onSuccess(info: Void?) {
```
Comment (bmarty) [inline matrix-sdk/src/main/java/org/matrix/androidsdk/features/integrationmanager/IntegrationManager.kt]: No need to wrap the callback if you do optimistic update

File: matrix-sdk/src/main/java/org/matrix/androidsdk/features/integrationmanager/IntegrationManager.kt [hunk 3]
```diff
@@ -106,17 +113,66 @@ class IntegrationManager(val mxSession: MXSession, val context: Context) {
...
+                AccountDataElement.ACCOUNT_DATA_TYPE_ALLOWED_WIDGETS,
+                updatedMap,
+                object : SimpleApiCallback<Void>(callback) {
+                    override fun onSuccess(info: Void?) {
+                        callback?.onSuccess(null)
+                    }
+                })
+    }
+
+    fun seVideoConfDomainAllowed(domain: String, allowed: Boolean, callback: ApiCallback<Void?>?) {
```
Comment (bmarty) [inline matrix-sdk/src/main/java/org/matrix/androidsdk/features/integrationmanager/IntegrationManager.kt]: Why generic "VideoConf" and after use hardcoded "jitsi"?

File: matrix-sdk/src/main/java/org/matrix/androidsdk/features/integrationmanager/IntegrationManager.kt [hunk 4]
```diff
@@ -106,17 +113,66 @@ class IntegrationManager(val mxSession: MXSession, val context: Context) {
...
+                    it[domain] = allowed
+                }
+
+        //Avoid to override delete unknwon keys
+        val updatedMap = (
+                mxSession.dataHandler.store
+                        ?.getAccountDataElement(AccountDataElement.ACCOUNT_DATA_TYPE_ALLOWED_WIDGETS)
+                        ?.content ?: HashMap()
+                ).apply {
+            set("jitsi", accountDataContent)
```
Comment (bmarty) [inline matrix-sdk/src/main/java/org/matrix/androidsdk/features/integrationmanager/IntegrationManager.kt]: "jitsi" should be pass in parameter "nativeWidgetName" or something like that
FTR nativeWidgetDomainPermission as map (key example: jitsi)

File: matrix-sdk/src/main/java/org/matrix/androidsdk/features/integrationmanager/IntegrationManager.kt [hunk 5]
```diff
@@ -106,17 +113,66 @@ class IntegrationManager(val mxSession: MXSession, val context: Context) {
...
+                videoConfDomainPermissions.filterNot { it.domain == domain } + listOf(DomainPermission(domain, allowed))
+        notifyListeners()
+
         mxSession.accountDataRestClient.setAccountData(
                 mxSession.myUserId,
                 AccountDataElement.ACCOUNT_DATA_TYPE_ALLOWED_WIDGETS,
-                mapOf("widgets" to accountDataContent),
+                updatedMap,
                 object : SimpleApiCallback<Void>(callback) {
                     override fun onSuccess(info: Void?) {
```
Comment (bmarty) [inline matrix-sdk/src/main/java/org/matrix/androidsdk/features/integrationmanager/IntegrationManager.kt]: Remove the wrapping

File: matrix-sdk/src/main/java/org/matrix/androidsdk/features/integrationmanager/IntegrationManager.kt [hunk 6]
```diff
@@ -68,8 +56,8 @@ class IntegrationManager(val mxSession: MXSession, val context: Context) {
     var integrationAllowed = true
         private set
 
-    private var widgetPermissions = emptyList<WidgetPermission>()
-    private var videoConfDomainPermissions = emptyList<DomainPermission>()
+    private var widgetPermissions = emptyMap<String, Boolean>()
+    private var nativeWidgetPermissions = emptyMap<String, Map<String, Boolean>>()
```
Comment (bmarty) [inline matrix-sdk/src/main/java/org/matrix/androidsdk/features/integrationmanager/IntegrationManager.kt]: Maybe document what are the keys and what are the values please

Comment (BillCarsonFr) [reply matrix-sdk/src/main/java/org/matrix/androidsdk/features/integrationmanager/IntegrationManager.kt]: done

File: matrix-sdk/src/main/java/org/matrix/androidsdk/features/integrationmanager/IntegrationManager.kt [hunk 7]
```diff
@@ -207,52 +163,43 @@ class IntegrationManager(val mxSession: MXSession, val context: Context) {
...
         override fun onStoreReady() {
             localSetIntegrationManagerConfig(retrieveIntegrationServerConfig(), false)
+            integrationAllowed = extractIntegrationProvisioning()
             widgetPermissions = extractWidgetPermissionFromAccountData()
+            nativeWidgetPermissions = extractNativeWidgetsPermissionFromAccountData()
             notifyListeners()
         }
     }
 
+    private fun extractIntegrationProvisioning(): Boolean {
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply matrix-sdk/src/main/java/org/matrix/androidsdk/features/integrationmanager/IntegrationManager.kt]: it's a private api, prefixed by extract as all other function that are extracting setting from account data

[the remark this answers is not in the document]
Comment (bmarty) [reply matrix-sdk/src/main/java/org/matrix/androidsdk/features/integrationmanager/IntegrationManager.kt]: ok, understood, thanks :)

**Pull request merged**